### ISSUE TITLE
Unable to export yolov8 model to tensorrt on Nvidia Agx Xavier 

### ISSUE THREAD

**Issue by Hasnain1997-ai** [ultralytics/ultralytics#1897]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hey!
I am trying to convert YoloV8 model(yolov8n) to tensorrt using the methods provided in the docs but I am unable to do so, though following the same method convert tensorrt model on 1080Ti GPU's system but not on Jetson Xavier.
Commond I am using is: **yolo export model=yolov8n.pt format=engine device='0'**

And the errors I am getting is:

**Ultralytics** YOLOv8.0.69 🚀 Python-3.8.10 torch-1.12.0a0+2c916ef.nv22.3 CUDA:0 (Xavier, 14898MiB)
YOLOv8n summary (fused): 168 layers, 3151904 parameters, 0 gradients, 8.7 GFLOPs

PyTorch: starting from yolov8n.pt with input shape (1, 3, 640, 640) BCHW and output shape(s) (1, 84, 8400) (6.2 MB)

ONNX: starting export with onnx 1.13.1 opset 14...
WARNING: The shape inference of prim::Constant type is missing, so it may result in wrong shape inference for the exported graph. Please consider adding it in symbolic function.
WARNING: The shape inference of prim::Constant type is missing, so it may result in wrong shape inference for the exported graph. Please consider adding it in symbolic function.
WARNING: The shape inference of prim::Constant type is missing, so it may result in wrong shape inference for the exported graph. Please consider adding it in symbolic function.
WARNING: The shape inference of prim::Constant type is missing, so it may result in wrong shape inference for the exported graph. Please consider adding it in symbolic function.
WARNING: The shape inference of prim::Constant type is missing, so it may result in wrong shape inference for the exported graph. Please consider adding it in symbolic function.
WARNING: The shape inference of prim::Constant type is missing, so it may result in wrong shape inference for the exported graph. Please consider adding it in symbolic function.
ONNX: export failure ❌ 6.0s: Expected all tensors to be on the same device, but found at least two devices, cpu and cuda:0! (when checking argument for argument other in method wrapper__equal)

TensorRT: starting export with TensorRT 8.5.2.2...
TensorRT: export failure ❌ 6.8s: expected str, bytes or os.PathLike object, not **NoneType**

I have tensorrt 8.5.2.2, Pytorch '1.12.0a0+2c916ef.nv22.3', Jetpack 5 Install(ubuntu20.04)

Can you help me how to solve the issue in Nvidia AGX Xavier
Thank you

**Comment by glenn-jocher** [comment]
@Hasnain1997-ai, it looks like the issue you are running into is due to TensorRT not being able to find the correct device to export to. Specifically, TensorRT is complaining about devices of type "cpu" and "cuda:0". To fix this, you should make sure that your TensorRT is properly configured to use the Xavier's hardware and that all libraries are installed correctly. Additionally, it may be possible that Jetpack 5 requires a different version of TensorRT than what you are currently using. So, you need to confirm if you have installed the correct version of all dependencies.

**Comment by Hasnain1997-ai** [comment]
@glenn-jocher  Thank you for your response!

I have TensorRt Installed correctly, can you list any checks that I need to run to find if TensorRt is correctly configured?

**Comment by glenn-jocher** [comment]
@Hasnain1997-ai since the export works on your 1080Ti but not on your edge device it's likely that a difference in the PyTorch, ONNX or TRT versions is causing the difference. We don't have any Jetson/Xavier devices locally to test, but if you want to debug you could always remove the Try Except Wrapper here on L439 on the ONNX and TRT export functions in exporter.py and try to debug this:

https://github.com/ultralytics/ultralytics/blob/c38b17a0d80958788d7fbf6b5159b756b7d56e07/ultralytics/yolo/engine/exporter.py#L439-L503

**Comment by Hasnain1997-ai** [comment]
OK I will look into it and will come back to you

**Comment by glenn-jocher** [comment]
@Hasnain1997-ai, to check if TensorRT is configured properly you can verify that all dependencies including CUDA, cuDNN, Python, and PyTorch are up to date and installed correctly. Additionally, you can verify that the TensorRT version you are using is compatible with Jetpack 5 Install (Ubuntu 20.04) and that the Xavier's hardware is properly configured. In general, issues with exporting a model to TensorRT on Jetson Xavier can be caused by differences in PyTorch, ONNX, or TensorRT versions, so ensure that all versions are in line with compatibility requirements.

**Comment by Hasnain1997-ai** [comment]
Thankyou @glenn-jocher!
  I will try out what you have mentioned and will get back to you then

**Comment by glenn-jocher** [comment]
You're welcome, @Hasnain1997-ai! Feel free to ask any further questions, and we will do our best to assist you. Remember to ensure that all dependencies are up to date and installed correctly, check TensorRT compatibility requirements with Jetpack 5 Install (Ubuntu 20.04), PyTorch, and ONNX versions. And also that Xavier's hardware is properly configured. Verify these steps to resolve the issue in exporting the YOLOv8 model.

**Comment by Hasnain1997-ai** [comment]
Hi @glenn-jocher 
Thank you for your time and guidance.

I resolved my issue.

Solution: Firstly I re-Installed Jetpack. Then I cloned your yoloV8 Git repo, and install requirements from requirements.txt( python3 -m pip install requirements.txt)

After that, I install pytorch and torch-vision from https://forums.developer.nvidia.com/t/pytorch-for-jetson/72048 for Nvidia Jetpack devices.

Then I uninstall torchvision installed through pip.

after that, I run the export yolov8 command and it runs successfully.

Thank you

**Comment by karunakr** [comment]
Hi @Hasnain1997-ai  

I could able to export the model to tensorrt. I am facing issue with runing inference. how are are you running inference using tensorrt model

**Comment by Hasnain1997-ai** [comment]
Hi @karunakr !

Great to hear that you can export your model successfully since it is the tricky part, if you can export it with out any issue then the inference should not be a big issue.

if you could have put the error you are facing then it would have be great and easy to help you solve the issue.

BTW inference using tensorRT is the same as that of the pytorch.pt model

`from ultralytics import YOLO
model_path= './train/weights/best.engine'
model = YOLO(model_path)

results = model.predict(source='video source', task='detect', stream=True, save=False, verbose=False, conf = 0.5)`

Feel free to reach out if the issue persist,

**Comment by karunakr** [comment]
Thanks a lot, @Hasnain1997-ai 
I can able to run inference. Thoughh without setting Half= True, i could see no improvement in performance, whereas with Half inference is 2X faster. Between, any idea about simplify and dynamic properties while exporting model

**Comment by glenn-jocher** [comment]
You're welcome, @karunakr! Regarding simplify and dynamic properties while exporting the model, simplify removes any unused parameters from the model to decrease the overall size of the exported model. Dynamic properties are used for methods to overrule a model's static properties at runtime. When exporting your model using the YOLOv8 repo, you can include the `simplify` parameter while exporting by setting it to True to simplify the model and reduce its size. If your model has dynamic properties, you can keep them while exporting by not using the simplify parameter. It's important to keep in mind that while simplify may decrease model size, removing some parameters can cause some degradation or compromise in the inference accuracy.

**Comment by karunakr** [comment]
Thanks @glenn-jocher I can see performance is the same with both simplify or without simplify. I can see 2x improvement when I added half to True, and the quality of the outcome is almost the same. 
As our input image size is the same (1280X720), I believe the dynamic property need not be set to True. I also configured imgsz=1280 while exporting the model. 

One last question related to this topic, what all dependencies do we need to add in the docker definition file to make tensorrt model to run.

**Comment by glenn-jocher** [comment]
Thanks for your update, @karunakr! I'm glad to hear that enabling half-precision increased the performance without a significant impact on the inferencing quality. Regarding the dynamic property, if the input image size remains the same, you don't require to enable dynamic property while exporting the model.

Regarding your last question, to run the TensorRT model in Docker, you would require all the necessary dependencies, particularly the dependencies that your TensorRT model uses. You should have all the required CUDA and CuDNN versions that your model requires. Ensure that the version of the developer libraries, runtime libraries, and samples are all consistent with the TensorRT version you're using. You should also ensure compatibility with the JetPack version you're using on the target platform. Once you have the dependencies installed properly, you can proceed with running the TensorRT model inside the Docker container. 

I hope that helps! Please feel free to ask any further queries you may have.

**Comment by karunakr** [comment]
Indeed it helps. Thanks @glenn-jocher for all your help

**Comment by glenn-jocher** [comment]
You're welcome, @karunakr! I'm happy to assist you with your queries. If you need any further assistance, please don't hesitate to reach out to the YOLOv8 community, and we'll be happy to help you out.

**Comment by shreejalt** [comment]
> Hi @glenn-jocher Thank you for your time and guidance.
> 
> I resolved my issue.
> 
> Solution: Firstly I re-Installed Jetpack. Then I cloned your yoloV8 Git repo, and install requirements from requirements.txt( python3 -m pip install requirements.txt)
> 
> After that, I install pytorch and torch-vision from https://forums.developer.nvidia.com/t/pytorch-for-jetson/72048 for Nvidia Jetpack devices.
> 
> Then I uninstall torchvision installed through pip. Also installed onnx-runtime-gpu from https://elinux.org/Jetson_Zoo after that, I run the export yolov8 command and it runs successfully.
> 
> Thank you

Hi, I am facing the same issue. I followed everythiong that you told. Can you please provide me the version of your onnxruntime, tensorrt and jetpack?

**Comment by glenn-jocher** [comment]
@shreejalt hi there! I'm glad to hear you've made progress. For the versions, it's important to ensure compatibility between ONNX Runtime, TensorRT, and JetPack. Typically, you'd want to use the latest stable versions that are compatible with each other. However, since I'm not the original author of the comment you're referring to, I don't have their specific version numbers. I recommend checking the NVIDIA documentation or forums for the compatible versions with your JetPack release. Good luck! 🚀
